### PR TITLE
[#310] OpenAPI v3: code assist for response code creates invalid value

### PR DESCRIPTION
**Pull request by ghillairet**
See #310 

This PR does not contain a test. I have to find a way to make the test run in standalone while loading the status-code.json file which is normally loaded through Activator.getBundle().getResource().

### CODE REVIEW

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            proposals.add(new Proposal(code, code, description, phase));
+        }
+
+        return proposals;
+    }
+
+    private Iterator<JsonNode> statusCodes(final String prefix) {
+        final boolean noPrefix = Strings.emptyToNull(prefix) == null;
+
+        return Iterators.filter(statusCodes.elements(), new Predicate<JsonNode>() {
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java]: We really should switch to Java 8...

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    public Collection<Proposal> getProposals(TypeDefinition type, AbstractNode node, String prefix) {
+        Collection<Proposal> proposals = Lists.newArrayList();
+
+        for (Iterator<JsonNode> it = statusCodes(prefix); it.hasNext();) {
+            JsonNode current = it.next();
+            String code = current.get("code").asText();
+            String description = current.get("description").asText();
+            String phase = current.get("phrase").asText();
+
+            proposals.add(new Proposal(code, code, description, phase));
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java]: Can we keep `displayString` as it is now and add a colon(`:`) to the end of the `replacementString` so user don't have to type it. For example:
* `displayString` = `200`
* `replacementString` = `200:`

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
+package com.reprezen.swagedit.openapi3.assist.ext;
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/ext/ResponseCodeContentAssistExt.java]: A missing copyright header?

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessor.java [hunk 4]
```diff
@@ -21,23 +22,25 @@
...
 import com.reprezen.swagedit.openapi3.assist.ext.SchemaTypeContentAssistExt;
 
 public class OpenApi3ContentAssistProcessor extends JsonContentAssistProcessor {
     
+    private static final JsonProposalProvider proposalProvider = new JsonProposalProvider(
+            new CallbacksContentAssistExt(), //
+            new SchemaTypeContentAssistExt(), //
+            new SchemaFormatContentAssistExt(), //
+            new ParameterInContentAssistExt(), //
+            new ResponseCodeContentAssistExt());
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessor.java:36]: Thanks!

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessorTest.xtend [hunk 5]
```diff
@@ -139,16 +146,37 @@ class OpenApi3ContentAssistProcessorTest {
...
+			  /foo:
+			    get:
+			      responses:
+			        <1>
+		''', document)
+
+		val proposals = test.apply(processor, "1")
+		assertThat(
+			proposals.map[(it as StyledCompletionProposal).replacementString],
+			hasItems("200:")
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessorTest.xtend]: I suggested to add a `:` to the `replacementString and it's expected in the `replacementString` in this test!

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessorTest.xtend [hunk 6]
```diff
@@ -139,16 +146,37 @@ class OpenApi3ContentAssistProcessorTest {
...
-		assertThat(proposals.map[(it as StyledCompletionProposal).replacementString],  
+		assertThat(
+			proposals.map[(it as StyledCompletionProposal).replacementString],
 			hasItems("int32", "int64", "float", "double", "byte", "binary", "date", "date-time", "password", "")
 		)
 	}
+
+// TODO
+//	@Test
+	def void testResponseStatusCode() {
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ContentAssistProcessorTest.xtend]: Our maven build runs tests as Eclipse plugin test. Let's re-enable this test so our CI could detect any regression.

**Comment by tfesenko** [comment]
Looks good, thanks!
Suggested improvements to be done before merging this PR :
- [ ] Add a copyright header
- [ ] Add a `:` to the replacementString
- [ ] Enable the new test

**Comment by ghillairet** [comment]
@tfesenko I updated the PR.

I added missing `default` and `x-` values. Also moved the ContentAssistExt class into the core project so that now the v2 editor can also benefit from it.

screen shot 2017-09-28 at 16 38 32

File: com.reprezen.swagedit.core/build.properties [hunk 7]
```diff
@@ -15,5 +15,6 @@ bin.includes = .,\
                lib/org.mozilla.rhino_1.0.0.7R4.jar,\
                lib/org.apache.commons.lang3_3.2.1.jar,\
                icons/,\
-               plugin.xml
+               plugin.xml,\
+               resources/
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/build.properties:19]: Thanks!

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        for (Iterator<JsonNode> it = statusCodes(prefix); it.hasNext();) {
+            JsonNode current = it.next();
+            String code = current.get("code").asText();
+            String description = current.get("description").asText();
+            String phase = current.get("phrase").asText();
+
+            proposals.add(new Proposal(code + ":", code, description, phase));
+        }
+
+        proposals.add(new Proposal("x-", "x-", null, "vendorExtension"));
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java:75]: Just checked the [spec](https://github.com/OAI/OpenAPI-Specification/blob/OpenAPI.next/versions/3.0.0.md#responses-object), confirming that the `responses` object can have vendor extensions.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import com.google.common.collect.Lists;
+import com.reprezen.swagedit.core.Activator;
+import com.reprezen.swagedit.core.assist.Proposal;
+import com.reprezen.swagedit.core.model.AbstractNode;
+import com.reprezen.swagedit.core.schema.TypeDefinition;
+
+public class ResponseCodeContentAssistExt implements ContentAssistExt {
+
+    private static final JsonPointer pointer = JsonPointer.compile("/definitions/responses");
+
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java:35]: It works for OAS3, but does not work for Swagger v2. I will create a separate issue for this.

**Comment by tfesenko** [comment]
Extracted #399 "Swagger v2: code assist for response code creates invalid value"
Also, the `replacementString` should whap the response code into single quotes `200:` -> `'200':`. It's an easy change. I will merge this PR now but keep the issue open with a comment about quotes.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java (hunk 9 above)
Comment (ghillairet) [reply com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/ext/ResponseCodeContentAssistExt.java:35]: Can you confirm it is not working for you for v2? I have it working for me.

screen shot 2017-10-04 at 16 15 19

**Pull request merged**